paddle/README.md

```diff
@@ -0,0 +1,141 @@
+In my mind, the memory package works like the following:
+
+## Design
+
+### Usage
+
+To allocate 4KB CPU memory:
+
+```cpp
+p = memory::Alloc(platform::CPUPlace(), 4*1024);
+```
+
+To allocate 4KB memory on the 3rd GPU:
+
+```cpp
+p = memory::Alloc(platform::GPUPlace(2), 4*1024);
+```
+
+To free memory and check the so-far used amount of memory on a place:
+
+```cpp
+auto pl = platform::GPUPlace(0);
+p = memory::Alloc(pl, 4*1024);
+cout << memory::Used(pl);
+memory::Free(pl, p);
+```
+
+### The API
+
+In `paddle/memory/memory.h` we have:
+
+```cpp
+template <typeanme Place> void* Alloc(Place, size_t);
```

Comment:
Actually, I think if we regulate the way we use GPUPlace and CPUPlace, like here -- two and only two specializations of `Alloc<Place>`, we might not need `typedef boost::variant<GPUPlace, CPUPlace> Place` at all. And this might help us remove the dependency to boost.